framework/Optimizers/Optimizer.py

```diff
@@ -92,44 +55,84 @@ def __init__(self):
-    self.inputInfo['SampledVars'     ]  = self.values               # this is the location where to get the values of the sampled variables
+    self.inputInfo['SampledVars']       = self.values               # this is the location where to get the values of the sampled variables
     self.constants                      = {}                        # dictionary of constants variables
-    self.FIXME                          = False                     # FIXME flag
     self.printTag                       = self.type                 # prefix for all prints (optimizer type)
-
-    self._endJobRunnable                = sys.maxsize               # max number of inputs creatable by the optimizer right after a job ends
-
+    self.submissionQueue                = {}                        # by traj, a place (deque) to store points that should be submitted some time after they are discovered
+    #functions and dataojbects
     self.constraintFunction             = None                      # External constraint function, could be not present
+    self.solutionExport                 = None                      #This is the data used to export the solution (it could also not be present)
     self.mdlEvalHist                    = None                      # Containing information of all model evaluation
     self.objSearchingROM                = None                      # ROM used internally for fast loss function evaluation
-
+    #multilevel
+    self.multilevel                     = False                     # indicates if operating in multilevel mode
+    self.mlBatches                      = {}                        # dict of {batchName:[list,of,vars]} that defines input subspaces
+    self.mlSequence                     = []                        # list of batch names that determines the order of convergence.  Last entry is converged most often and fastest (innermost loop).
+    self.mlDepth                        = {}                        # {traj: #} index of current recursion depth within self.mlSequence, must be initialized to None
+    self.mlStaticValues                 = {}                        # by traj, dictionary of static values for variables in fullOptVars but not in optVars due to multilevel
+    self.mlActiveSpaceSteps             = {}                        # by traj, integer to track iterations performed in optimizing the current, active subspace
+    self.mlBatchInfo                    = {}                        # by batch, by traj, info includes 'lastStepSize','gradientHistory','recommendToGain'
+    #stateful tracking
     self.nextActionNeeded               = (None,None)               # tool for localStillReady to inform localGenerateInput on the next action needed
-
+    self.status                         = {}                        # by trajectory, ("string-based status", arbitrary, other, entries)
```

Comment:
After a small benchmark on this, even using Python 3.6's enum implementation is slower than a string comparison for the status' given below.

```python
import time
import random
import numpy as np

## For Python 3:
from enum import Enum
class Interaction(Enum):
    No = 0
    Yes = 1
    Debug = 2
    Test = 3

## For Python 2:
# class Interaction:
#     No, Yes, Debug, Test = range(4)

problemSize = int(1e7)
numTrials = 10

indices = list(Interaction)
ints = list(range(4))
strings = ['No No No No No No No No No No',
           'Yes Yes Yes Yes Yes Yes Yes Yes',
           'Debug Debug Debug Debug Debug',
           'Test Test Test Test Test Test']

randEnums   = [random.choice(indices) for _ in range(problemSize)]
randInts    = [random.choice(ints)    for _ in range(problemSize)]
randStrings = [random.choice(strings) for _ in range(problemSize)]

################################################################################
times = np.zeros(numTrials)
for j in range(numTrials):
    counts = {'No':0,'Yes':0, 'Debug':0, 'Test':0}
    start = time.time()
    for i in range(problemSize):
        number = randInts[i]
        if number == 0:
            counts['No'] += 1
        elif number == 1:
            counts['Yes'] += 1
        elif number  == 2:
            counts['Debug'] += 1
        elif number  == 3:
            counts['Test'] += 1
    end = time.time()
    times[j] = end-start

print('integer: {} +/- {} s'.format(np.mean(times), np.std(times)))

################################################################################
times = np.zeros(numTrials)
for j in range(numTrials):
    counts = {'No':0,'Yes':0, 'Debug':0, 'Test':0}
    start = time.time()
    for i in range(problemSize):
        number = randStrings[i]
        if number == 'No No No No No No No No No No':
            counts['No'] += 1
        elif number == 'Yes Yes Yes Yes Yes Yes Yes Yes':
            counts['Yes'] += 1
        elif number == 'Debug Debug Debug Debug Debug':
            counts['Debug'] += 1
        elif number == 'Test Test Test Test Test Test':
            counts['Test'] += 1
    end = time.time()
    times[j] = end-start

print('string: {} +/- {} s'.format(np.mean(times), np.std(times)))

################################################################################
times = np.zeros(numTrials)
for j in range(numTrials):
    counts = {'No':0,'Yes':0, 'Debug':0, 'Test':0}
    start = time.time()
    for i in range(problemSize):
        number = randEnums[i]
        if number is Interaction.No:
            counts['No'] += 1
        elif number is Interaction.Yes:
            counts['Yes'] += 1
        elif number is Interaction.Debug:
            counts['Debug'] += 1
        elif number is Interaction.Test:
            counts['Test'] += 1
    end = time.time()
    times[j] = end-start

print('enum: {} +/- {} s'.format(np.mean(times), np.std(times)))
```

integer: 3.047193694114685 +/- 0.09608294677338775 s
string: 3.0560888767242433 +/- 0.07831414372368399 s
enum: 6.504127144813538 +/- 0.16359815605516845 s